mypy/ipc.py

```diff
@@ -0,0 +1,216 @@
+                # this empty write is to copy the behavior of socket.sendall,
+                # which also sends an empty message to signify it is done writing
+                _winapi.WriteFile(self.connection, b'')
+            except WindowsError as e:
+                raise IPCException("Failed to write with error: {}".format(e.winerror))
+        else:
+            self.connection.sendall(data)
+            self.connection.shutdown(socket.SHUT_WR)
+
+    def close(self) -> None:
+        if sys.platform == 'win32':
+            if self.connection != _winapi.NULL:
+                _winapi.CloseHandle(self.connection)
+        else:
+            self.connection.close()
+
+
+class IPCClient(IPCBase):
+    """The client side of an IPC connection."""
+
+    def __init__(self, name: str, timeout: Optional[int]) -> None:
+        super().__init__(name)
+        if sys.platform == 'win32':
+            timeout = timeout or 1000  # we need to set a timeout
+            try:
+                _winapi.WaitNamedPipe(self.name, timeout)
+            except FileNotFoundError:
+                raise IPCException("The NamedPipe at {} was not found.".format(self.name))
+            except WindowsError as e:
+                if e.winerror == _winapi.ERROR_SEM_TIMEOUT:
+                    raise IPCException("Timed out waiting for connection.")
+                else:
+                    raise
+            try:
+                self.connection = _winapi.CreateFile(
+                    self.name,
+                    _winapi.GENERIC_READ | _winapi.GENERIC_WRITE,
+                    0,
+                    _winapi.NULL,
+                    _winapi.OPEN_EXISTING,
+                    0,
+                    _winapi.NULL,
+                )
+            except WindowsError as e:
+                if e.winerror == _winapi.ERROR_PIPE_BUSY:
+                    raise IPCException("The connection is busy.")
+                else:
+                    raise
+            _winapi.SetNamedPipeHandleState(self.connection,
+                                            _winapi.PIPE_READMODE_MESSAGE,
+                                            None,
+                                            None)
+        else:
+            self.connection = socket.socket(socket.AF_UNIX)
+            self.connection.settimeout(timeout)
+            self.connection.connect(name)
+
+    def __enter__(self) -> 'IPCClient':
+        return self
+
+    def __exit__(self,
+                 exc_ty: 'Optional[Type[BaseException]]' = None,
+                 exc_val: Optional[BaseException] = None,
+                 exc_tb: Optional[TracebackType] = None,
+                 ) -> bool:
+        self.close()
+        return False
+
+
+class IPCServer(IPCBase):
+
+    BUFFER_SIZE = 2**16
+
+    def __init__(self, name: str, timeout: Optional[int] = None) -> None:
+        if sys.platform == 'win32':
+            name = r'\\.\pipe\{}-{}.pipe'.format(name, os.getpid())
```

Comment:
Instead of using the pid, it might be better to use some (base64d?) bytes out of `os.urandom`? It doesn't *really* matter for our uses but makes it a nicer bit of library code, at least, since then one process can have multiple servers.

Reply:
Yeah, good point. I've thought about the idea of breaking out the ipc module into its own library, I'm sure there are others who might be interested in this code.